File: next/src/shared/vaultTreeExpand.ts
export type VaultTreeExpandMode = 'expandAll' | 'collapseAll'

export type VaultTreeLikeNode = {
  kind: 'file' | 'directory'
  relativePath: string
  children?: VaultTreeLikeNode[]
}

export function isVaultTreeExpandMode(value: unknown): value is VaultTreeExpandMode {
  return value === 'expandAll' || value === 'collapseAll'
}

export function collectDirectoryPaths(nodes: VaultTreeLikeNode[]): string[] {
  return nodes.flatMap((node) => (
    node.kind === 'directory'
      ? [node.relativePath, ...collectDirectoryPaths(node.children ?? [])]
      : []
  ))
}

export function ancestorDirectoryPaths(relativePath: string): string[] {
  if (!relativePath.includes('/')) return []
  const parts = relativePath.split('/')
  const paths: string[] = []
  for (let index = 1; index < parts.length; index += 1) {
    paths.push(parts.slice(0, index).join('/'))
  }
  return paths
}

export function resolveExpandedPaths(
  directories: Iterable<string>,
  mode: VaultTreeExpandMode,
  selectedPath = ''
): Set<string> {
  if (mode === 'expandAll') return new Set(directories)
  return new Set(selectedPath ? ancestorDirectoryPaths(selectedPath) : [])
}

export function ensureAncestorsExpanded(expanded: ReadonlySet<string>, selectedPath: string): Set<string> {
  if (!selectedPath) return new Set(expanded)
  const next = new Set(expanded)
  for (const path of ancestorDirectoryPaths(selectedPath)) next.add(path)
  return next
}

File: next-compose/composeApp/src/desktopMain/kotlin/com/rememberber/mootool/next/compose/ui/components/VaultTreeExpand.kt
package com.rememberber.mootool.next.compose.ui.components

import com.rememberber.mootool.next.compose.storage.VaultEntry

/** 对照 Electron `vaultTreeExpand.ts` `ancestorDirectoryPaths`。 */
fun vaultAncestorDirectoryPaths(relativePath: String): List<String> {
    if (!relativePath.contains('/')) return emptyList()
    val parts = relativePath.split('/').filter { it.isNotEmpty() }
    return (1 until parts.size).map { index -> parts.take(index).joinToString("/") }
}

/**
 * 对照 Electron `resolveExpandedPaths`：`collapseAll` 时仅保留 [selectedPath] 的祖先目录展开，
 * 避免选中深层文件后「全部折叠」把当前项藏进折叠树。
 */
fun vaultTreeExpandForMode(
    expandMode: String,
    items: List<VaultEntry>,
    selectedPath: String = "",
): Map<String, Boolean> {
    val directories = items.filter { it.directory }.map { it.relativePath }
    return when (expandMode) {
        "collapseAll" -> {
            val keep = vaultAncestorDirectoryPaths(selectedPath).toSet()
            directories.associateWith { it in keep }
        }
        "expandAll" -> directories.associateWith { true }
        else -> directories.associateWith { path -> !path.contains('/') }
    }
}
